expo/expo
# Expo In App Purchases

# Why

This PR adds the [highly requested](https://expo.canny.io/feature-requests/p/in-app-purchases) In-App Purchases module to allow Expo apps to be monetized.

# How

Using the [Google Play Billing](https://developer.android.com/google/play/billing/billing_library_overview) library on Android and [Storekit](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/StoreKitGuide/Introduction.html#//apple_ref/doc/uid/TP40008267-CH1-SW1) API on iOS.

I had to make some changes in order to get started including adding `expo-in-app-purchases` as a dependency to Sandbox, adding the EXInAppPurchases pod as a dependency in the podfile, and changing the Bundle Identifier of the Expo Client from `host.exp.Exponent` to `dev.expo.Payments`on iOS and the Application ID from `host.exp.exponent` to `com.expo.payments` on Android. This was necessary to match the entries in the Google Play Console and App Store Connect respectively so that I can safely add purchases to test against without putting the Expo client at risk of being removed. 

This is because in order to test In App Purchases on both platforms you have to create an entry for the app and add the purchases yourself along with the necessary details (title, price, description, etc). On Android, you also have to upload an apk and publish the app as an alpha or beta release.

Obviously, we do not want the changes to be permanent so those commits **HAVE TO BE REVERTED** before merging this PR into master. This was just the fastest and easiest way to start prototyping.

# Test Plan

Currently, I am building the Expo client from source (after making the aforementioned changes) and testing using the sandbox app under the `apps/` directory as I'm developing.

In the future, we will need a more reliable way to test this feature and we can do one of the following:
1. Change the Application ID/Bundle Identifier of Standalone NCL to match the above, reupload them to Google Play Console/App Store Connect, and add a screen that uses the API.
2. Add a new standalone app in the `apps/` directory with the above App/Bundle IDs.
3. Keep using sandbox or another simple Expo app and add instructions in the README for how to get started testing such as changing the Expo clients App IDs (obviously this is less ideal).

As of now, you should be able to just pull down this branch, build the client, and test the API in sandbox.

Please note that in order to test the Google Play Billing and Storekit libraries you have to be using a real device so this will not work on simulators.
